### ISSUE TITLE
Discussion on the future of staging

### ISSUE THREAD

**Issue by rderooy** [dosbox-staging/dosbox-staging#217]
Hi,

Considering what happened on Vogons, I think a discussion is needed on the future of dosbox-staging.

The way I see it, these are the options:

1) Reconciliation. Talk to the DOSBox developers and make it work, and get them to unblock your Vogons account. Personally I don't think this is realistic.

2) Quit your dosbox development. I hope you won't, but it is probably what some would like.

3) Rename dosbox-staging to something else, as obviously the name is no longer applicable, and deviate freely from upstream dosbox and try to get a larger user base. 

4) Merge with another fork that does not care too much about upstream. Personally I think it would be great if you could work on dosbox-x. I know that in the past you mentioned that it was too focused on demo's, and while at first that was the case, it has not really been lately. Also, those demos are good cases to stress dosbox in often unexpected ways. Perhaps with your help dosbox-x could gain enough traction, such that it can be added to the Linux distributions as an alternative.

Both 3) and 4) Will undoubtedly require a new place to hold discussions as your Vogons account is blocked and any new threads can be easily locked or deleted.

**Comment by es20490446e** [comment]
Better trying to reach consensus first. But don't waste your time with abrasive people, they never change.

I would stay with the official DosBox project if they made a clear statement themselves about how they will do it from now own, without hidden excuses.

Otherwise don't count of them, they would just continue doing the same. In that case better to have a talk with DosBox-X and see if the differences of scope could be solved.

**Comment by es20490446e** [comment]
Anybody new coming to this topic, you can see the previous conversation [here](https://github.com/dreamer/dosbox-staging/issues/206).

**Comment by DominusExult** [comment]
1. again, I'm not a Dosbox developer but I was part of previous discussions and my input might help.
Reconcile but you may have to do/accept some things:
- don't advertise staging over Dosbox. Meaning: "you have a problem with Dosbox, come to our github, file the issue and we discuss it there!" feels very disrespectful. "You have a problem with Dosbox, why don't you write it in their forum. Or alternatively here is our fork and we can discuss it there." sounds way better. When people come to Vogons and are not sure where the official source/place to discuss is, it's clear that there is a problem.

- don't badmouth Dosbox developers. It's okay to say you have differences in what you consider important. But half truths like in the linux config file issue don't help a strained relationship.

- accept that you have different goals and scope than upstream. If upstream says it's important to keep compatibility with certain OS' than don't bring it up again and again. If that means upping the c++ standard is not acceptable, bringing up patches that up it, feel like a waste of time.

- accept that version 0.75 is a goal and certain things like SDL2 are planned after that to minimize the regression fallout. 

- accept that eliminating compiler warnings are not the goal of the devs. Especially if it's not sure that changes to that will slow down Dosbox.

- again, don't vent your frustration over these issues online. Of course it's alright If you write you see it different but your posts on Vogons were often close to being nasty about this.

Again, I'm not Qbix and this is my opinion

**Comment by krcroft** [comment]
@rderooy, thanks for creating this thread to discuss what's transpired at Vogons in relation to how this affects dosbox-staging going forward.

Indeed, Vogon's staff froze the dosbox-staging thread and banned @dreamer's account. These actions were carried out by the Vogon's administrators making decisions based on their internal governance model, aspects of which are used to govern the open-source DOSBox project given the Vogons administrators also happen to be core developers of DOSBox.

The difference in how DOSBox _is_ governed (as an open-source C++ project) versus how @dreamer feels it _should_ be governed is what drove the creation of dosbox-staging in the first place.

So to me, I see everyone at Vogons and here sharing the same end-goals: maintain the ability to play DOS games on modern hardware!, but with merely different approaches to how this project should be governed.  @dreamer has captured some of these in the `README.md` document, and even more clarity on how dosbox-staging will be governed will come in the way of the **Contributors Guide** and **Code of Conduct**.

Based on the above, I don't think the Vogon's actions require changes to `dosbox-staging` given these actions stem from fundamentals already clearly understood by @dreamer. 

We also need to remember that governance models have little bearing on the exchange of fungible end products.  Two oil companies might be managed very differently but the resulting oil commodity can be exchanged without issue.  So too it is with high-quality open-source code, and its exchange will continue between projects without concern.

**Comment by es20490446e** [comment]
@krcroft I think governance is the most important aspect of a project, as it sets if people would be able to collaborate or not.

Lack of time is just the symptom of not being able to make use of the time of others, usually because of not trusting anybody to be able to come with their part without watching them all the time.

The need of control means that the human organisation isn't good enough to work well by itself. Aka don't manage the people, manage the system.

**Comment by joncampbell123** [comment]
DOSBox-X developer here...

DOSBox-X focuses more on general emulation and accuracy, rather than just games as the main project does. That does come at a cost of a bit of emulation performance of course (also pointed out by a comparison video from Pixelmusement), but it helps make the emulation more accurate.

I tend to focus on the demoscene (especially anything prior to 1996) only because that era of the MS-DOS scene tends to have all manner of weird hardware tricks, bugs, and speed-sensitive issues that make them the perfect kind of stuff to test emulation accuracy against, even moreso than old DOS games. But that's not all I test against.

I also test against DOS games and applications: https://github.com/joncampbell123/msdostest

Lately, due to the PC-98 emulation mode added in late 2017, that includes PC-98 games and all the weird things they do with the hardware and the internals of the DOS kernel.

I also test to make sure that it can run Windows 95 through ME, which other forks cannot do because of the recursive manner that DOSBox handles page faults, vs preemptive multitasking environments like Windows 95 which apparently allow a task switch during a page fault, and issues with the normal CPU core when interrupted like that.

If the intent of this fork is just games, then it would be better to take a different fork and work from that, with perhaps patches from DOSBox-X where emulation accuracy may help.

**Comment by dreamer** [comment]
*I am banned from Vogons by IP and don't have even read access there, so it will be a bit harder to provide links to specific posts in there - I will do it in future edit to this post.*

> 1. Reconciliation. Talk to the DOSBox developers and make it work, and get them to unblock your Vogons account. Personally I don't think this is realistic.

Me neither. Contrary to @DominusExult's narrative, we already had a consensus reached with Qbix via IRC discussion - it was reached on February 29th, and Dominus nor DosFreak were **NOT** present on IRC when it was reached - but krcroft was (so their repeated calls for "talk to the devs" are just noise - we did). The result of that consensus was:

- no migration to Git will take place, upstream will stay in SVN, we stopped suggesting it
- no new features will be accepted, we stopped submitting such patches
- no C++11 code will be accepted, we stopped submitting such patches
- no code cleanups nor refactorizations will be accepted, we stopped submitting such patches
- evident bugfixes for crashes and similar bugs are acceptable, as long as it's C++03
- fixes for improving emulation for specific games are acceptable, as long as it's C++03

We stopped pushing for C++11 even earlier, as we were waiting for *any* response to our questions between December 9th and February 29th.

Based on these parameters, when a user asked me to bypass upstream project and send patches directly to Debian package maintainers - my response was: NO. And when asked for clarification about no new features landing upstream, I provided it as well.

And I feel rather vindicated considering my assessment was correct: respecting XDG spec was asked for at least 3 times between 2011 and 2020 and in all cases, users were ignored or ridiculed.

All the development I do is in open and using GPL-compatible licensing - if DOSBox authors want to use any of the code developed for dosbox-staging, they are free to do so - but I am not going to spend any more time worrying about what is or isn't acceptable upstream and will not engage in any discussions on Vogons or DOSBox Sourceforge project.

> 2. Quit your dosbox development. I hope you won't, but it is probably what some would like.

Not planning on doing so - the software does not work correctly for my use cases yet, and there's plenty of stuff to fix. I don't plan on doing everything myself, so I will keep doing what I am doing: making the development process as streamlined as possible, fixing code, engaging DOS gaming communities to reach new users and developers.

> 3. Rename dosbox-staging to something else, as obviously the name is no longer applicable, and deviate freely from upstream dosbox and try to get a larger user base.

When creating the repo I spent a lot of time thinking about the naming - decided to go with "-staging" suffix to communicate similar relation as wine-staging has towards Wine or linux-staging towards Linux. Well, the nuance got lost on the Vogons crew… I *really, really* did not want to create a fork - just needed SDL2 integrated without waiting another 10 years.

We are halfway towards the first stable release, so renaming now is really inconvenient… but it's just another curveball thrown our way.

I don't want to rush with project rename - there are many nuances about good naming (it affects the creation of logo/new icons, affects program name, affects configuration file locations, needs to be SEO friendly, etc, etc). This is a topic for a whole new discussion.

>  4. Merge with another fork that does not care too much about upstream. Personally I think it would be great if you could work on dosbox-x. I know that in the past you mentioned that it was too focused on demo's, and while at first that was the case, it has not really been lately. Also, those demos are good cases to stress dosbox in often unexpected ways. Perhaps with your help dosbox-x could gain enough traction, such that it can be added to the Linux distributions as an alternative.

I never said DOSBox-X is "too focused" on demoscene! :) It is their focus, and it's reflected in their GUI, plentiful configuration options and features they prioritize to work on.

From a developer perspective, there are quite large differences between DOSBox-X and dosbox-staging - my main reason not to join that project was their plan to remove SDL dependency completely (which makes sense for a project focused on non-interactive demoscene!), while I want to integrate more closely with SDL2 (as it lowers the long-term maintenance cost).

I will approach maintainer of DOSBox-X though and ask about his opinions and plans (just not today).


@DominusExult said:
> * again, don't vent your frustration over these issues online. Of course it's alright If you write you see it different but your posts on Vogons were often close to being nasty about this.

I think users can read [the closed thread](https://www.vogons.org/viewtopic.php?p=790065#p790065) and decide for themselves who was being nasty in there. Literally starting from your first response to my post.

**Comment by joncampbell123** [comment]
@dreamer DOSBox-X as implemented now supports both SDL and SDL2. It can be done, though some functionality is lost including 3Dfx OpenGL mapping and the Direct3D output in Windows. However DOSBox-X uses it's own internal copy of SDL which has been heavily modified over the years to fix bugs and support cases it was never originally intended to handle.

I also have a fork of DOSBox-X called dosbox-rewrite, which is currently a DOSBox-X stripped of SDL (so it's SDL2 only), and a lot of non-essential emulation removed, in an attempt to clean up and rewrite a better architecture. dosbox-rewrite will have a better name later on. One area of focus is the VGA emulation, which currently works but feels like hacks and tweaks piled on hacks and tweaks and could be better implemented with a cleanup.

**Comment by DominusExult** [comment]
> hmm, seems more like you want to move things away from the DOSBox devs as fast as  possible...

^^^ I don't see anything nasty in this first reply of mine. In hindsight this and my next reply were more prophetic, wasn't it?
But if you really don't see how you badmouthed the Dosbox devs here on github, reddit,... then... well, ok. Great for you.

I was well aware of your consensus with Qbix, so, yes, I probably did not add anything new in this regard, my points were more a rehash.
The important part of my points was the badmouthing and redirecting everything directly to staging.

**Comment by dreamer** [comment]
@joncampbell123 
Me and @krcroft went a bit different route, but end result is probably similar to dosbox-rewrite (I will take a look - thank you for mentioning it!) - we removed all the SDL1 cruft from the code and made sure the project builds and links cleanly with libraries provided by various repositories (Linux repos, vcpkg on Windows, brew on macOS). When using up-to-date SDL2 we had no reason to patch it in any way so far.

This way the code is cleaner, build process faster and easier to automate on CI. I remember reading in the issue about organizing work for DOSBox-X "next-gen" that SDL is to be dropped completely, but maybe I misunderstood or the plan changed? When dealing with Demoscene there's no reason to support things like new game controllers, so it seemed reasonable to me.

> some functionality is lost including 3Dfx OpenGL mapping

Yeah, properly integrating that patch into dosbox codebase will need a lot of work still, and I suspect an alternative implementation from scratch might be necessary. I tried using this functionality in the past and had game-breaking issues in all the titles I tested (sometimes small ones like no video cutscenes, sometimes big ones like missing textures or no video output at all). I arrived at the conclusion, that SDL2 has much higher priority.

> and the Direct3D output 

You meant DirectDraw? We removed it completely, and rely on SDL2 texture API to handle all platform-specific renderers (`direct3d`, `directx11`, `metal`, `opengles`, `opengles2` - AmigaOS users even mentioned they have their own platform-specific API). We just let SDL2 do its thing and it generally works; we have some issues with `metal` and `directx11`, but there's WIP fix prepared already.

Personally, I think the best move forward would be to remove `output=surface` (software driver for `output=texture` should work just as well, but all that old code for sustaining software-only rendering and scaling could be removed) and then start work on implementing Vulkan renderer (with the goal of eventually deprecating OpenGL in few years).

> One area of focus is the VGA emulation, which currently works but feels like hacks and tweaks piled on hacks and tweaks and could be better implemented with a cleanup.

VGA emulation code is really hard to comprehend right now…

Our analysis indicated that old OPL emulation stuff is in really bad shape (many subtle off-by-1 errors, buffer overruns, etc - but we switched to Nuked as default, so at least it does not affect users).

I also have one nasty crash in CPU emulation code, that I can reproduce ~80% of the time (but reproduction steps take around 2 minutes to reach the crashing part, grrr).

Anyways,

I will take a closer look at dosbox-rewrite repo and contact you in the following week - thank you for contacting us :)

**Comment by joncampbell123** [comment]
I mean Direct3D. DOSBox-X inherited the Direct3D output from Daum's branch. It's a hack that uses SDL to set up the window, and then slap Direct3D9 output on top of the window. It's a hack that SDL2 can take the place of.

**Comment by es20490446e** [comment]
Let me know when you have reached a conclusion, so I can package it 👍

**Comment by sikthehedgehog** [comment]
I'm not a DOSBox-X contributor (though I'm subscribed to its issue tracker), but um, looks like people here still get the impression that it's oriented towards running demos?

As far as I can tell the focus is better accuracy/compatibility. Demos saw a lot of attention early on because they were obviously broken (and exposed a lot of issues that affected everything, and from what I remember it also exposed a lot of gaps in sound emulation and a bunch of FPU issues too?), though most of the reported issues I see nowadays seem to be about broken games instead (e.g. a recent one was games crashing when VESA BIOS reported a lot of resolutions because they allocated a fixed buffer to hold the video modes and it didn't fit… that needed adding a setting to limit the number of reported modes as a workaround).

**Comment by joncampbell123** [comment]
@sikthehedgehog Correct, DOSBox-X is more than just demos, though running through the 1986-1995 demoscene was a great way to find all the accuracy issues. The idea is for more accurate emulation, enough to make retro development possible with a reasonable expectation that real hardware works mostly the same. Another idea is to add configuration options so you can test your code by playing "what if?", so there are options to turn things on and off to make it possible.

**Comment by joncampbell123** [comment]
@sikthehedgehog Quake is one good example: It was written to allocate a buffer, query a video mode into it, and free it. It looks like it was later modified to return early if the mode was not 256-color or did not fit other criteria, and for those cases, forgot to free the buffer. So a lot of non 256-color VESA modes will cause a lot of leakage with Quake and Quake will eventually query a VESA BIOS mode to a NULL pointer and trash the interrupt vector table (it allocates without checking for a failure to allocate).

**Comment by joncampbell123** [comment]
I want to clarify dosbox-rewrite is a fork of DOSBox-X with all but the most essential core code ripped out so that anyone including myself can do a clean rewrite of the emulation from that.

Also, minimum C++ standards for each project:
DOSBox SVN - C++03 (as I understand it)
DOSBox-X - C++11
dosbox-rewrite - C++14

**Comment by es20490446e** [comment]
Perhaps if the code is modular enough you won't need two versions.

I have always abused modularity and this has been great when refactoring code, or trying to understand it. Even the functions are highly modular, being as short as possible.

It sounds quite basic, but it has big consequences.

**Comment by dreamer** [comment]
About minumum language standards:

**DOSBox SVN** - C++03 - based on: [this thread](https://sourceforge.net/p/dosbox/patches/283/)
**dosbox-staging** - C++11 right now, but the last technical issue that prevented us from moving to C++14 is resolved already, we just haven't needed anything specific from C++14 yet; C++17 is not going to happen - at least for few more years (unfortunately; the reason being ABI compatibility); Rust might happen.

Based on the comments here and looking up the source code of DOSBox-X and dosbox-rewrite: I think my previous assertion about very different goals of our projects stands.

dosbox-staging is not concerned about perfect emulation. This is not a design goal for us - we'll take extra steps to ensure games are playable (especially in respect of various available user options), but rather won't go an extra mile to emulate original hardware behaviour. Perfect emulation is a worthy and applaudable design goal, just not *our* design goal.

We care about:

- first-time user experience and ease of use (also clearly motivated new features)
- developer experience and changes geared towards creating positive-feedback loops (clean code, nice commit messages, wide-reaching CI and automation, static code analysis, we're working towards test suite of automatic regression tests, unit testing, etc); we're also taking extra care in recording development history, authorship and source of the changes, and motivation behind them (this is already paying dividends)
- making the project more maintainable (no hacks, no side-loading build dependencies, no fighting against OS, making life easy for distributions and distributors, etc)
- clearly motivated incremental improvements, not revolutions

To accomplish these goals we're not only adding and modernizing code, but also removing harmful and unnecessary one and deprecating/removing/fixing user options that cause confusion without real benefit.

So, based on these goals, at the moment I don't see how we could *easily* merge the projects (not in short-term perspective at least). I do have few ideas for cooperation though (some more radical, some less) - I will need a few days to crystalize them in an email form.

**Comment by joncampbell123** [comment]
Exactly. Imagine the arguing and fighting with developers I could have caused trying to merge all the changes I made upstream for the last 8 years... I could have easily gotten myself kicked off of the Vogons forums :) That's why DOSBox-X is a fork that diverges from the main project.

However the accuracy and numerous options of DOSBox-X could help determine why something is failing in the main project, and I'm happy to point out what it might be from experience. Other patches may be useful, such as the 8086 and 286 cores and the "reset vector" thing needed for Windows 3.x to run in 286 "standard" protected mode.

Is there any interest in PC-98 emulation as well for your version?

I work on DOSBox-X as a hobby and I don't mind if this or other forks borrow code from it as long as you don't claim it as your own, feel free to do so, just as I take care to credit the source when I borrow code into DOSBox-X.

Another thing to point out is that DOSBox-X and DOSLIB tend to develop in parallel. DOSLIB is a library of various MS-DOS stuff, compiled with NASM and Open Watcom, that I've written to determine actual hardware behavior as well as demonstrate how to program MS-DOS and hardware in general that's been in development since it started in 2008. Feel free to use source and compiled binaries from DOSLIB to test your emulation.

**Comment by es20490446e** [comment]
Does accurate hardware emulation, as the DosBox-X does, add much complexity to the software?

**Comment by joncampbell123** [comment]
A fair bit, yes. Another loss of performance is the additional use of PIC_AddEvent() and PIC_FullIndex() to track time and schedule things accurately.

The dual IBM/PC-98 handling complicates the DOS kernel a bit as well, which is why I would rather make a separate fork for those who want FM Towns emulation than complicate the code further.

The accuracy is important though to run a wider variety of software including time sensitive stuff, and to help give retro DOS developers a better idea what actual hardware is like.

Your version, like SVN if I understand, would be interested only in what it takes to run DOS games and accuracy outside that scope is not needed. However the accuracy and options in DOSBox-X would probably assist your project in understanding what is going wrong.

One option that might be helpful to add would be the vmemdelay option. Some older DOS games don't page flip properly on faster hardware, or they redraw the screen in a manner that only works on slow video memory. Some examples off the top of my head are Catacombs 3D and Corncob 3D. Both need vmemdelay to avoid on screen flickering.

**Comment by joncampbell123** [comment]
Actually based on this weird joke game unearthed by Shovelware Diggers, there's a patch that probably should be made for DOSBox SVN and your fork: The "old VBE" emulation should not zero fill the entire 256-byte area when querying the VESA BIOS because whatever video library or framework used in "Get Saddam" uses a too-small buffer to query the VESA BIOS if it detects it. I'm guessing that in 1991 when the game was made, VESA BIOSes only wrote out the structure as far as it needed to, so it was OK, but zeroing out 256 bytes overwrites important pointers that follow that area of memory and the game will crash from it (which is why machine=vgaonly solves the issue).

**Comment by dreamer** [comment]
@es20490446e There's no easy answer to this question; simple stats generated using tokei:

- DOSBox SVN codebase is 123K LOC (122K of C/C++)
- dosbox-staging is ~164K LOC  (~160K of C/C++). *We removed lots of stuff, but also e.g. cleaned up and ported sdl_sound, and added header-only codec libraries for CD-DA, which are quite large (34K LOC).*
- DOSBox-X is ~1065K LOC (~855K of C/C++) - *large chunk of this probably comes from bundled libraries, but it's still 5x times as much code*

SVN already covers emulation of more than ~90% of all DOS games (maybe more than 95%), and probably >99% of culturally significant ones.

Also, there's a lot of code in DOSBox-X, that we **don't** want (like OS/2 support, physical CD audio playback, menu-based GUI) and stuff we do have already is missing (built-in codecs, shell improvements, CI covering multiple platforms, pixel-perfect mode, accurate emulation of CD-DA via compressed tracks), and some stuff we do want to work on is missing as well.

I know there is stuff in DOSBox-X we also **do** want (fixed PC speaker emulation is the first one, that comes to mind), but most of them are low priority to us.

@es20490446e if you are trying to decide which project to contribute to, or which project to use as basis for your own fork - you will need to look into the code and decide for yourself.

**Comment by es20490446e** [comment]
I will test 👍

**Comment by es20490446e** [comment]
Some interesting tests:
- [DosBox compared with DosBox-X](https://youtu.be/J4x02RICPBI)
- [DosBox-X compared with DosBox-Staging](https://youtu.be/_xDIDd0RuLI)

And:
- [Config file used for staging](https://bin.privacytools.io/?e155450cd48357ba#MlOyt4u9b5zIXvPeXsl3DSZWtShVU1Pj+Bn7jb15FwY=)
- [Zone Raiders download](https://archive.org/download/msdos_Zone_Raiders_1995/Zone_Raiders_1995.zip)

**Comment by es20490446e** [comment]
Also interesting discoveries on [tests](https://youtu.be/rPPqMT080So) with [Realms of the Haunting](https://archive.org/details/RealmsOfTheHaunting).

**Comment by krcroft** [comment]
Alberto; very nice play testing of the core DOSBox project vs DOSBox-X.  I like your suggestion to @joncampbell123 regarding finding a way to save UI customizations to the local running game folder (perhaps as a minimal delta that ride on top of the defaults). 

One clarification is that DOSBox-0.74.3 and DOSBox SVN are the same "core" project by the team at vogons.org.  You can refer to these both of these as "DOSBox" or "official DOSBox"; and then either the latest release, 0.74.3, or SVN as hosted by sourceforge.

dosbox-staging, this project here on github, is a soft-fork of DOSBox-SVN (and tracks SVN tightly).  The latest dosbox-staging binaries are artifacts as built by our CI chain (click 'Actions' in the menu, then Linux, then assets in the top right).  We have a recently optimized Linux build here: 
https://github.com/dreamer/dosbox-staging/suites/518113500/artifacts/2858923

Thanks for the easy to absorb results! 
Video is worth a thousand pictures :-)

**Comment by joncampbell123** [comment]
Whoah, is that an FPU emulation issue with normal core?

https://www.youtube.com/watch?v=J4x02RICPBI&feature=youtu.be

DOSBox SVN and MSVC builds of DOSBox-X truncate floating point registers to the 64-bit "double type" because Microsoft's compiler does not support "long double". The only exception to that rule is if you use dynamic core.

MinGW builds of DOSBox-X use long double and should never have that problem, normal core or otherwise.

**Comment by joncampbell123** [comment]
I'm also noticing the absurd percentages showing in the title bar. That obviously needs to be fixed.

**Comment by dreamer** [comment]
@es20490446e What version you used for testing? Because I suspect you downloaded source code of pre-release tagged version (which is functionally equivalent to SVN r4258). This is an easy mistake to make, as we haven't finished our first stable release yet :) 

I suspect you used an older version, because:

1. staging generates .conf file named `dosbox-staging-git.conf`, not `dosbox-SVN.conf`. We did some changes to available options, defaults, and documentation in that file - that's why I am bringing it up.
2. You mentioned a bug when pressing 'up' - this is exactly the same problem I had when playing MegaRace 2 and Tomb Raider 1; move to SDL2 fixed it (happened ~December)
3. There is `SET BLASTER…` line displayed in the shell (right after blue 'intro' screen of DOSBox) - we made this line hidden long time ago.
4. Other small details (e.g. `aspect=true` by default)

You can run `git describe` command in the cloned repo to display somewhat human-readable version description. 

You can also download and run our [snapshot build](https://github.com/dreamer/dosbox-staging#development-snapshot-builds) ([link straight to artifacts section of latest build](https://github.com/dreamer/dosbox-staging/actions/runs/57262099)) - they are carefully crafted to work on any modern-ish Linux distro.

If you'll decide to build it yourself, then I strongly recommend using `-O3` flag for any performance comparison; DOSBox code (especially in CPU emulation area) depends on compiler optimizations very much. @krcroft is working on autoFDO optimised builds, but it was not merged to master yet.

As I said earlier - testing the wrong version is an easy mistake to make, as we haven't tagged any of development versions after that. Hopefully, you'll see the dropped 'up' key fixed :)

**Comment by joncampbell123** [comment]
Regarding PC speaker changes, DOSBox-X's changes reflect some realities about the 8254 PIT.

Writing the mode byte resets the phase of the square wave, for one, which is what gives PC speaker output by some games. Duke Nukum does it's sound effects by writing the PIT counter and mode byte at a fixed rate (about 10-20Hz I think) whether the frequency changes or not. The change in mode byte at that rate is what gives the sound effects the rougher "edge" on real hardware compared to DOSBox SVN.

Many accuracy changes in DOSBox-X are made by using the FillUp() function in the mixer code to render audio up to that point in time before making a state change, including PC speaker, Adlib, Tandy and Sound Blaster emulation. That simple change allows correct rendition of rapid state changes.

- The guitar digitized sample in the title screen of Skate or Die in Tandy mode renders correctly through the Tandy 3-voice with that change (instead of a muddy buzz)
- Any game that makes rapid changes to the OPL FM chip to make "digitized speech" through pure Adlib support will work that way.

The "goldplay mode" requires more modifications than that to work, since it requires Sound Blaster emulation to fetch DMA at the programmed rate instead of at the per-millisecond interval it normally does, to catch each byte written by the game/demo's ISR, but it can be done. "goldplay mode" is easy to detect: The DMA channel is initialized with a transfer length of 1 or 2 bytes with auto-init and it's always an 8-bit channel, and the block length sent to the DSP is often much larger. DOSLIB programming experiments shows that you can do that with 16-bit PCM on SB16, but what's the point?

An unusual test case for your staging build that might be of interest is the 1992 demo "Crystal Dream" which has very unusual Sound Blaster support code. Not only does it use the "goldplay mode" of sample-by-sample ISR rendering, but it also uses an unusual method to keep the DSP playing via DMA. Evidently written prior to DSP 2.x commands that allow "auto-init" mode, and apparently written to avoid the need for an IRQ handler, the demo relies on the fact that when SB and SB Pro hardware is playing from DMA the busy bit of the DSP (port 22Ch) rapidly cycles on it's own accord. The rapid cycling stops when the DSP finishes playing audio. The ISR polls port 22Ch to detect this, and issue another DSP play command (command 14h) to keep audio going when it stops cycling. SB16 hardware is always cycling the busy bit for whatever reason, which prevents the demo from detecting that condition and is the reason the demo's Sound Blaster support fails on SB16 hardware. Another quirk revealed by the demo is that you can apparently issue DSP command 14h to play, and before the block completes, issue another DSP command 14h to effectively restart the block (as far as the DSP is concerned). So if the demo is not exactly on-time about issuing DSP command 14h, or it seems to trigger early, issuing another DSP command 14h too early doesn't hurt anything.

It might be a fun test case, though as far as I know no DOS game out there ever does that with Sound Blaster.

**Comment by joncampbell123** [comment]
If you want to support DMA emulation that supports decrement mode, a good test case is the EMFINTRO demo. It's Sound Blaster support code likes to render normally to DMA, but backwards, and it programs the DMA channel in decrement mode to match (which is why in DOSBox SVN you get random noise and garbage instead because it only supports increment mode).

The demo also relies on the fact that on SB and SBPro hardware port 22Dh is an alias of port 22Ch (it will hang otherwise). Look in the debugger when it hangs, you'll notice reading from port 22Dh.

**Comment by joncampbell123** [comment]
@es20490446e Do you get better game performance if you set vmemdelay=0. DOSBox-X by default emulates some video memory delay because video memory was traditionally slower than system memory when access by the CPU.

**Comment by grapeli** [comment]
@es20490446e 
Zone Raiders in low resolution looks bad (very). For this game I recommend you [dosemu2](https://github.com/stsp/dosemu2). It is currently a virtual machine with KVM support.
In terms of performance, it beats all versions of the dosbox. **Raw Power**.
Zone Raiders in 640x480 flies like crazy at ten years of intel westmere (2.4GHz).
A small comparison:
Quake 1 800x600, timedemo demo1
-[dosbox-staging with AutoFDO optimizations](https://github.com/dreamer/dosbox-staging/suites/518113500/artifacts/2858923) 32.6 FPS
-dosemu2  198.2 FPS

[dosemu2 ubuntu PPA](https://code.launchpad.net/~dosemu2/+archive/ubuntu/ppa)

**Comment by i30817** [comment]
Personally and as the person that accidentally started this minor explosion i think everyone is over reacting. I suggested debian because debian already patches dosbox regularly and it's in their wheelhouse to accept distro-standards patches, and yes, because dosbox does not accept feature patches in my experience.

Dreamer is persistent and Vogons moderators zealous and dosbox devs semi-inaction, and the fact that the project doesn't move that fast was always going to cause friction. It's still a bad idea not to communicate, even if you don't want to collaborate, especially since no 'real' insults flew, just 'dosbox doesn't want patches' and 'you sound like to fork dosbox' (self fulfilling prophecy here).

Don't get into a situation where trolls cause a 'rivalry' for kicks. Trolls are always bad even if they're 'supporting you'.

Anyway i'm not innocent of blowing up on Vogons for dosbox stuff. For complete transparency and as verification i'm not one of the 'trolls' i cautioned above here are my beefs with dosbox-as-a-project:

I despised - but didn't express it on Vogons - dosbox rolling over for Roland threats and making the mt32 emulation experience much worse by not building with munt as a library, but that is understandable and munt saved the day anyway. 
Recently i posted a venting thread about dosbox still using svn causing me a bit of work and a switch to a git mirror that is not 'upstream' (so i can't really be certain nothing is introduced there). This worries but not enough to give up, so i vented unnecessarily and a Vogons mod closed the thread right away. I still think still using svn now is terrible but whatever.

**Comment by es20490446e** [comment]
I will have a look at your points.

**Comment by dreamer** [comment]
@i30817 

> Recently i posted a venting thread about dosbox still using svn causing me a bit of work and a switch to a git mirror that is not 'upstream' (so i can't really be certain nothing is introduced there). 

Branches and tags named `svn/*` in dosbox-staging repo are reproducible mirrors of SVN. If you want to make sure no additional stuff was injected in there, review and run script: `scripts/import-from-svn/import-from-svn.sh` (it takes ~50 minutes to dump full history from SVN - script will create new local git repo on your machine).

Respective commit hashes in your newly created local repo and hashes in dosbox-staging should match. If they'll ever won't match - create a bug report, please.

**Comment by i30817** [comment]
I might do that in the future when the build breaks again (inevitable). Currently i'm using launchpad autoimport of https://github.com/jwilk-mirrors/dosbox and the guy keeps dozens of mirrors and keeps svn import signatures, so i'm sort of covered legally if something happens.

mirror of a mirror of svn...

**Comment by bluddy** [comment]
This reminds me very much of the vim vs neovim debate. vim was basically a dead product under closed development. The author was unwilling to accept very useful patches of any kind. Eventually, a user created neovim, with cleaned up code (they removed support for non-existent hardware and OSs, for example) and a new vision for the architecture. Amazingly, once the competition of neovim existed, vim suddenly decided to copy every one of the major features introduced by neovim (built-in terminal, asynchronous jobs interface) and the original, dead project suddenly became alive again.

I recommend calling staging neodosbox, or dosbox-neo, as a reference to that example, and moving right along. Make releases. dosbox can choose to compete or die.

**Comment by dreamer** [comment]
@bluddy very apt comparison - and I would say, vim project was even more welcoming towards new contributors and contributions than DOSBox is right now:

- vim moved away from SVN in 2010 (now development happens primarily in Mercurial with officially maintained Git repo), DOSBox devs are determined to stay in SVN
- vim had clear rules for communicating patches via a mailing list, unlike DOSBox, where rules for submitting patches are not clear at all

I am happy user of both vim (for quick edits) and neovim (for heavier stuff) right now, as both have their strengths.

An idea of naming the project "neobox" to mimic "neovim" did cross my mind…

**Comment by es20490446e** [comment]
Keep the name short and cool 😎

**Comment by joncampbell123** [comment]
Testing DOSBox staging... it looks DOSBox-X like functionality was added where moving the mouse over the window matches the cursor in DOS games. Works great. :+1:

**Comment by joncampbell123** [comment]
Is anyone working on machine=ega and aspect ratio calculations?

machine=ega, aspect=true, output=opengl gives me a 640x350 screen that is way too wide for what EGA was actually like. EGA monitors were nothing like the widescreen displays we have now.
![egawidescreen](https://user-images.githubusercontent.com/6245486/77100170-70847900-69d2-11ea-8aee-d6ca09635de3.png)

**Comment by joncampbell123** [comment]
Here's another one: DOSBox-X maps host cursor position to the window by the min/max of the INT 33h interface, however it will clamp to the nearest multiple of the screen dimensions if close enough because some games like Daggerfall set the maximum to width/height of the screen minus about 16 to 32 pixels. My guess is that doing that allows the game to draw the mouse cursor without having to worry about clipping against the edge of the screen.

The DOSBox-X adjustment allows the guest cursor to match host cursor even in Daggerfall.

Anyone here willing to merge that into this project from DOSBox-X?

**Comment by krcroft** [comment]
This would be a huge improvement @joncampbell123 ; which right now is essentially broken for games that perform custom positioning of the cursor (typically the SVGA games).  Thanks for figuring out what's going on under the hood, and the generous offer to provide a patch.

**Comment by dreamer** [comment]
Discussion about broken ega mode to be continued in: #228

@joncampbell123 
Regarding cursor: improvements compared to SVN are largely thanks to up-to-date SDL2 - but it's not enough - some games work great (e.g. X-COM), others almost perfect (Larry 7 - movement is good, cursor is deterministically shifted), other really bad (Worms - different acceleration on both axis, similar to Win 3.11 behaviour); when in fullscreen, the behaviour is ok across the board.

But we do have changed default and new user options, e.g. mouse can be locked to window using middle-mouse button, which is much more convenient than default ctrl-f10 (can be disabled if game needs MMB). Personally I consider DOSBox-X cursor behaviour to be the best of all DOSBox versions so far :) any tips on how to get there will be greatly appreciated (or PRs ;)).

This is an interesting problem to work on, because people who test DOSBox often, get used to the mouse issues and learn to live with it, but new users consistently have problems with mouse lock/mouse capture.

**Comment by joncampbell123** [comment]
Some games ignore mouse position and use only the relative motion reported by INT 33h or through the interrupt handler.

One trick DOSBox-X does is, when the mouse is not captured, absolute cursor position is applied and relative motion is not counted. If the mouse is captured, then absolute cursor position is ignored and relative motion is counted and sent.

If a game is only listening for relative motion, then moving the mouse over the emulator window should do nothing.

**Comment by joncampbell123** [comment]
As for Windows 3.1, DOSBox-X has an "integration device" at I/O ports 28h-2Bh (IBM PC mode) or DB28h-DB2Bh (PC-98 mode). There is a Windows mouse driver in the DOSLIB project that you can install that reads the absolute mouse position from DOSBox-X to give you "guest pointer integration" VirtualBox style.

**Comment by es20490446e** [comment]
That said `aspect=true` gives me the wrong aspect ratio, where `false` gives me the right one.

**Comment by dreamer** [comment]
@es20490446e The right one? I think the *right* aspect ratio is 4:3 - `aspect=true` should trigger scaling to make this happen. `aspect=false` keeps square pixels and results in 1.83:1.

In my opinion, `aspect` option is seriously misnamed, it should be named `square_pixels=true/false` or something similar :(

**Comment by joncampbell123** [comment]
As far as I understand it, aspect=true is supposed to stretch the display to a 4:3 aspect ratio like a monitor would. Part of the calculation is factoring in the active display region to determine the final display region.

aspect=false is just rendering the pixels as 1x1 square on your display which adds it's own distortion for anything other than 640x480.

**Comment by krcroft** [comment]
Roughly 5% of well known games used art that assumed square pixels. No doubt the majority of these were ports from other platforms, and the DOS team wasn't able to adapt the art - so kept it as is. 

@es20490446e, here is my list of `aspect = false` games (while roughly 600 others require the 4:3 vertical stretch or `true`):

- Afterlife
- Battle Chess Enhanced
- Beneath a Steel Sky
- Discworld
- Disney's The Jungle Book
- Disney's The Lion King
- Ecstatica II
- Elite I Plus
- Flight Unlimited
- Hocus Pocus
- Jazz Jackrabbit
- Jazz Jackrabbit Holiday Hare 1994
- Jazz Jackrabbit Holiday Hare 1995
- Lemmings 3D
- Lure of the Temptress
- Master of Orion 2
- MAX Mechanized Assault & Exploration
- Overlord
- Prehistorik 1
- Scorched Earth
- Seek and Destroy
- System Shock
- Terminator 2 The Arcade Game
- The Incredible Machine 1
- The Incredible Machine 2
- Warlords II Deluxe
- Worms
- Z

**Comment by dreamer** [comment]
I would add UFO: Enemy Unknown and X-COM: Terror from the Deep to this list, but it might be a personal preference ;) in these games strategic part of the game has art in 16:10 while tactical map is in 4:3; Jazz Jacrabbit experiences similar disconnect I think…

Perhaps one day we'll have a feature to set memory watcher to trigger square pixels on/off based on memory state of the emulated game… perhaps ;)

**Comment by es20490446e** [comment]
Could you give some examples of games that are better rendered with `aspect=true`?

**Comment by krcroft** [comment]
> Could you give some examples of games that are better rendered with `aspect=true`?

**Comment by dreamer** [comment]
This discussion reminded me of another thing I want to fix: #229 :)

In that list supplied by @krcroft: over last few months I tested several titles from that list, in all cases 4:3 is the correct aspect ratio (e.g. Alone in the Dark series looks almost grotesque when stretched via square pixels).

**Comment by es20490446e** [comment]
[On aspect ratio](https://youtu.be/zsUPzv__O_4)

**Comment by es20490446e** [comment]
By default NVIDIA graphics stretch the screen, not preserves the aspect ratio. Which can be set up in the NVIDIA control panel.

You can also test if the aspect ratio is the right one with exactly these settings:
```
fullresolution = desktop
windowresolution = desktop
output = openglnb
aspect = false
scaler = advmame3x
```

As it can be that your screen hardware scaler is incapable to keep aspect ratio at the lowest resolutions, but do so on higher ones. Resolutions like, for example, the one of the desktop.

Or you can test if ratio is preserved with and old Windows game, with the lowest resolution set.

As I said my screen keeps the right aspect ratio no matter what, without `aspect = false`. I purposely looked into this when buying the equipment, to be able to play old games without workarounds.

**Comment by dreamer** [comment]
@es20490446e You are incorrect; note how blue circle covers the knob neatly in the bottom freeze-frame of your video, while on upper frame, red pixels of the knob are visible to the left and right of the blue circle.

![circles](https://user-images.githubusercontent.com/3967/77189217-8039b580-6ad7-11ea-879c-b9cbbc43e871.png)

This is much more visible when circular element is bigger than few pixels apart; your usage of mame scaler also masks it a little bit. Here's much more visible example:

![image](https://user-images.githubusercontent.com/3967/77189686-3ef5d580-6ad8-11ea-98ef-d738381135d9.png)

Taken from [this excellent explanation of the problem](https://www.gamasutra.com/blogs/FelipePepe/20150423/241730/No_MSDOS_games_werent_widescreen_Tips_on_correcting_aspect_ratio.php).

My guess is: you just got used to incorrect aspect ratio for DOS games, that's all :). Detailed explanation is in the link @krcroft provided you earlier.

**Comment by krcroft** [comment]
Aspect ratio problem /really/ jump out at me, and I cannot bear to watch incorrect retro youtube channels. I've asked myself why are others less sensitive to it?

The pattern seems to be that, if you grew up in the 70s/80s using CRTs for your formative years - then aspect ratio is very obvious to you. If you grew up in 90s and early 2000's then this was a transition era from CRT to LCD, and was plagued with stretched signals from many sources that assumed 4:3 output included TV broadcasts, DVD players, and video game consoles. This later generation is very tolerant of the "stretched" look.. and indeed, to them, it is perfectly normal.

There have been arguments about _what the original artists intended_, given we cannot go back in time and ask them.  Therefore, the truth can be ambiguous.  In some cases though, the pixel art reflects actual physical models, making (at least those examples) unambiguous:

![Aspect_ratio_comparison_cyberdemon](https://user-images.githubusercontent.com/1557255/77193554-b0636300-6a9b-11ea-9088-bab6ce0daa2f.jpg)

**Comment by es20490446e** [comment]
Damn 🤣

I guess people get confused because the stretching is not that obvious, as it doesn't fill the screen just the resolution.

The image is just slightly flattered, two black bars are still at the sides, and the image looks sharper when not aspect corrected. So people would wrongly assume that is already aspect corrected somehow.

It could even be that, despite the slight stretching, the non corrected image is more comfortable after all. As it's sharper and fills more area of a modern screen.

**Comment by es20490446e** [comment]
[Dosemu2 test](https://youtu.be/KRLm4mKhUVU)

**Comment by es20490446e** [comment]
[Latest staging build just crashes](https://youtu.be/meiBISoD0kE)

**Comment by krcroft** [comment]
> [Latest staging build just crashes](https://youtu.be/meiBISoD0kE)

This is a known problem in general, https://www.dosbox.com/comp_list.php?showID=897&letter=Z
You can fix it with these steps:

1. Replace Zone Raiders' DOS4GW protected-mode handler inside RAID.EXE using DOS32A's SB program. Inside DOSBox, run `sb /r raid.exe`, which will modify the executable in place. Next time you run RAID.EXE, you will notice DOS32/a initializing instead of DOS4GW.

2. Use a local `dosbox.conf` file inside Zone Raiders, with at least these settings:

```
[dosbox]
machine  = vesa_nolfb
memsize  = 24

[cpu]
core      = dynamic
cycles    = 30000

[dos]
xms            = true
ems            = false
umb            = true
```

Even at 30,000 cycles (which isn't much), the game runs smooth.

**Comment by grapeli** [comment]
> [Dosemu2 is slower than DosBox](https://youtu.be/KRLm4mKhUVU)

Find the five differences between dosbox-staging and dosemu2 in this video.
https://drive.google.com/file/d/1VWo8i_X_Gig0X022e1U7jJpnBj4NWeJf/view?usp=sharing

**Comment by dreamer** [comment]
So now I need to do my video to show-off, sigh :P https://www.youtube.com/watch?v=anDNl0JFHi4 (at the end I started to mess with going in and out of fullscreen ;)).

There's FPS counter in top-right; it's a bit lower than normal due to OBS recording and running under X11 (when running via XWayland without OBS I have constant 70FPS). DOSBox has max limit of 70FPS and I haven't figured out why exactly yet.

Running it inside Steam via [Boxtron](https://github.com/dreamer/boxtron) (to get FPS counter). I followed **krcroft**'s advice for setting up the game, except `cpu` section; Boxtron automatically picks following settings:

```
[cpu]
core = auto
cputype = auto
cycles = max 95%
```
… because when playing the game, you want to play the game and not keep your CPU at 10% utilization with `cycles = 30000`, right? ;)

Testing v0.75.0-pre-686-ga5ce551e (my working branch: `po/gfx-splash-6`), compiled with -O3.

Anyway, what I wanted to say here… this game is perfectly playable using dosbox-staging (but DOS32/A stuff is unfortunate! I hope we can do something about it).

I am sure **dosemu*** is a great project, that serves users very well (especially when it comes to professional emulation of old DOS business and industrial software). @grapeli - if it works for you, then great - keep using it. But it has nothing to do with the future of dosbox-staging (which is topic of this discussion, despite offtops ;) ) and I see no area for cooperation here - both projects have practically nothing in common.

**Comment by es20490446e** [comment]
dosemu2 performs better at higher resolutions, but it's choppy at any.

Same as comparing a Windows virtual machine with Wine, I feel the latest will always win.

**Comment by es20490446e** [comment]
I'm seeing no difference with `cycles = max 95%`.

**Comment by volkertb** [comment]
> I am sure **dosemu*** is a great project, that serves users very well (especially when it comes to professional emulation of old DOS business and industrial software). @grapeli - if it works for you, then great - keep using it. But it has nothing to do with the future of dosbox-staging (which is topic of this discussion, despite offtops ;) ) and I see no area for cooperation here - both projects have practically nothing in common.

Awww. :disappointed:

Now that dosbox-staging was likely to become a complete fork anyway, I was actually about to open a feature request here for an optional hypervisor-backed CPU core that would leverage hardware-assisted virtualization (if available) instead of emulating a CPU. After all, on an x86-based machine, how could complete CPU emulation ever be faster than virtualization, with only stufff like VGA and sound cards having to be emulated? (And in the case of VESA mode, even the former could be paravirtualized, leaving basically only the sound card to be emulated in software.)

Granted, very old DOS games would actually run too fast when running natively on the host CPU, but particularly for beefier DOS extender games from the mid to late '90s, running them virtualized would result in greater performance, right?

I remember the DOSBox developers on Vogons being vehemently opposed to such a feature, even as just an optional one. But if you can run code natively on the host CPU, why not do so? That would always be more efficient, right?

As a matter of fact, why not go a step further and see if dosbox-staging and dosemu2 could merge?

Just to be clear, I'm merely proposing an **optional** CPU core here, not to throw out CPU emulation outright. Use emulation on different architectures (ARM, RISC-V), and use virtualization when the host CPU architecture happens to be compatible with the guest architecture. Why not?

Regardless of the above discussion, though: I'd like to express my appreciation for your work on dosbox-staging, @dreamer. It's really unfortunate how things didn't work out between you and the upstream DOSBox developers. Although DOSBox has been an amazing piece of software, the core developers seem to have developed a bit of an ivory tower mentality over the years, which, sorry to say, has lead to an apparent stagnation in the development of that project. It's good to see so many patches from the community finally coming together somewhere now.

Also, yay for your willingness to consider accepting Rust code in this project. :smiley:

**Comment by es20490446e** [comment]
> If you can run code natively on the host CPU, why not do so?

Show me a video of Dosemu running Zone Raiders with smooth performance, cause I suspect virtualization cannot accomplish that on consumer grade processors.

**Comment by dreamer** [comment]
@volkertb said:
> (…) I was actually about to open a feature request here for an optional hypervisor-backed CPU core that would leverage hardware-assisted virtualization (if available) instead of emulating a CPU. (…)

Go ahead and open such feature request then :) I do not plan to work on it (my backlog is overflowing and I haven't brought dosbox code to state when I can start work on features I *do* plan to work on yet), but the problem space warrants to be explored ;)

I will slap "help wanted" label on it, and people who want to work on it can experiment, discuss, and coordinate in there. If eventually a new feature will be developed, which won't break existing stuff, will adhere to our quality standards (clean, working, maintainable code!), and *will be easy to use*, then I will be ok with merging it in.

> Granted, very old DOS games would actually run too fast when running natively on the host CPU, but particularly for beefier DOS extender games from the mid to late '90s, running them virtualized would result in greater performance, right?

Some games (old, but I think some newer ones as well) already do work too fast in emulated CPU, and even my 8-year old laptop with integrated GPU is powerful enough to emulate any game from 90s at comfortable speed.

In my experience, the DOSBox emulation is already fast enough and accurate enough (not perfect! just *good enough*). It's just darn hard to configure the thing and work around all the bugs and quirks and inconsistencies - so that's the problem I am trying to tackle now.

> As a matter of fact, why not go a step further and see if dosbox-staging and dosemu2 could merge?

I think there's no common base nor vision to make such move feasible.

Projects, where I see some potential for the merger are: ECE and @duganchen's [fork](https://github.com/duganchen/dosbox).

With @joncampbell123's DOSBox-X - I think we are going to cooperate without merging the projects.

In the past, I approached developers of:
- [DOSBox-LFN](http://www.columbia.edu/~em36/wpdos/dosbox-vdos-lfn.html) - maintainer respectfully declined, as dosbox-staging does not overlap with his goals
- [libretro DOSBox core](https://git.retromods.org/pub/dosbox-svn) - maintainer saw no area for cooperation (this was a bit surprising to me actually)

There are dozens or more other DOSBox forks or projects incorporating DOSBox code out there, but I don't know any other maintained one.

Personally, I am considering merging this project with my other one, [Boxtron](https://github.com/dreamer/boxtron), but I think it's too early ATM; I am also working on features, that would make it possible for this project to be incorporated as a [Proton submodule](https://github.com/ValveSoftware/Proton/issues/1566), but it's a long shot and there's very high probability nothing will come out of it.

> Also, yay for your willingness to consider accepting Rust code in this project. smiley

I am not in "Rewrite It In Rust" crowd, but when I look at DOSBox code, I get "oh, in Rust I wouldn't have *this* problem" moment at least once a day.

Right now, DOSBox code is almost entirely single-threaded, and I found areas where it's starting to show certain limitations - I dread forcing multi-threaded implementation into a single-thread code (been there, done that)… Rust would be a huge help here. Some other technical wins would be: better standard library, trivial access to SIMD instructions, unit tests built-in the language, `cargo` and all the goodies, that come with it… There are some drawbacks as well, unfortunately.

Projects like librsvg or Firefox proved, that careful migration is possible. However, right now I am not working on moving to Rust (so far I only investigated it and experimented a bit) - the priority is 0.75.0 release. It is one of the topics I intend to tackle later on.

**Comment by bluddy** [comment]
> I am not in "Rewrite It In Rust" crowd, but when I look at DOSBox code, I get "oh, in Rust I wouldn't have this problem" moment at least once a day.

I would strongly caution against a Rust rewrite. Rewrites are doable with huge teams of people all working on the same task. Trying to do this kind of thing with a small team is doomed to failure. It's far more natural to slowly shift DOSBox to C++, as you're doing, all while keeping a fully working product.

Regarding multithreading, it's a very difficult task. DOSBox is a virtual machine (not the real kind, the theoretical kind) for instructions that all assume they're running in the same thread (since multithreading wasn't available in DOS). Unless you use multithreading to handle some hardware functionality, such as a specialized processor (ie. MT-32/Adlib or the like -- the PC doesn't have too many candidates), which can run in parallel, you really can't do very efficient multithreading. Multithreading could be useful for emulating x86 on non-x86 platforms, for example.

**Comment by dreamer** [comment]
@bluddy That's why I am not talking about rewrite! I see Rust would be beneficial for non-emulation code (shell, sdlmain, filesystem, support functions and libraries). In emulation code I see only very, very few, very specific cases where SIMD will be beneficial (e.g. speeding up NukedOPL because right now it's too resource-intensive for RaspberryPis - and I don't fancy writing Neon instructions by hand).

**Comment by JoshuaFern** [comment]
Reiterating here a previous conversation we had on Discord about renaming the project, I think we particularly liked "dreambox" but it would be valuable to get some more naming ideas and feedback.

Also don't take this as a confirmation that the project will be renamed, that's not up to me :)

**Comment by rderooy** [comment]
DreamBox is already a company name. Probably not a good idea.

**Comment by JoshuaFern** [comment]
@rderooy Good point, we should avoid picking existing trademarked names particularly when it touches the same industry.

**Comment by duganchen** [comment]
> Projects, where I see some potential for the merger are: ECE and @duganchen's fork.

I was wondering why this thread showed up in my inbox...

I didn't yet read it all, but IMHO the main thing that my fork has to offer you is the OpenGL 3 renderer, which is written by me. It does need SDL2, which you're already using. I have a single patch for that, which should be easy to adapt:

https://www.vogons.org/viewtopic.php?p=568163#p568163

It's patch #2.

For patch #1, which is meant to be applied first, I took the "SDL2" patch that someone posted to Vogons (everyone saw that, right?) and I stripped out all the Android stuff. If you did your own port to SDL2, keep that in mind.

Anyway, take whatever you want.

**Comment by dreamer** [comment]
@duganchen I did use the same SDL2 patch as a base (but only after replacing SDL_sound dependency to keep CD-DA emulation working, with @krcroft's help), also stripped all Android stuff (Android port is desirable, but out of scope for now), and I'm now in the process of cleaning it up and modernizing to prevent various bugs and issues.

But this fork is not just a collection of patches :) The question is: would you be interested in joining as co-maintainer and switching to work on a single repo, instead of maintaining a separate codebase - you would be joining me, @krcroft (author of modernized CD-DA functionality), and @ant-222 (author of pixel-perfect functionality) - we accept PRs from drive-by collaborators as well, of course ;)

I think our forks have very similar goals but seriously differ in details - including workflow, automation, features we decided to prioritize, and strategy of merging-in changes from upstream (I'm not sure - is your fork based on SVN trunk or SVN 0_74 branch?).

I don't know if you plan to continue development on your fork - note, that your future sync-up with SVN will be awkward, as upstream implemented GLSL shaders as well now (we have it merged-in already).

If you are not interested - I will add your OpenGL 3 renderer patch to the queue of things we are working on :) However, if you are open for collaboration - we can discuss details (also in private, if you find it more comfortable - my email address is attached to commits in the repo).

**Comment by duganchen** [comment]
@dreamer No thank you.

**Comment by dreamer** [comment]
@duganchen That's fine, thank you for response and details about your patch :) Cheers!

**Comment by es20490446e** [comment]
I suggested the name Bitbox, which means:
- Games are so low res you can see the "bits" on your screen (aka the "box").
- Ended as "box" to suggest its relationship with Dosbox.
- Pronounced like "beatbox", cause this got cooler.

Just a funky suggestion.

**Comment by bluddy** [comment]
If we're in the name suggestion game, I'd throw in 'unobox'. This plays on the fact that 'dos' is 2 in spanish. 'Tres' would have been better (dosbox++) but doesn't work as well.
Also, now that I think about it, `dosbox++` might be best.

**Comment by dreamer** [comment]
**About naming, I must put a disclaimer**: I find it's too late to change the name before 0.75.0 release (and I need a stable release for a number of reasons), so rebranding is for next release after that (at least).

Rebranding is a serious effort and it would push back the release by another month or so (I think I spent at least a week of my free time on converting raster DOSBox logo to vector format, going through multiple revisions before arriving at current variant, developing and testing scripts for icon conversions for Windows and macOS, developing and testing low-resolution icon variants, debugging icons (icons on Windows can be buggy, imagine that!). Also, documentation, manual, and buildsystem would need to be adjusted.

Changing command name alone will cause serious repercussions limiting our reach to other projects and users (Wine, Lutris, GameHub, MiniGalaxy, eXo; even for me it would cause more work to adapt Boxtron), AND hamper our development of regression tests on top  of that… There are also ergonomic and non-technical aspects to consider (e.g. is there already a Github project with such name that would rank higher in Google?)

*(breathe in)*

I want to rebrand, but it's not going to be spurious off-the-cuff decision :)

"Bitbox" - I like it, but there are several projects using this name already (some Bitcoin exchange, and a DIY console, probably more)

"DOSBox++" (with potential for our own command `dosboxpp`, which could be aliased to `dosbox` via Linux update-alternative mechanism) - I like this idea.

**Comment by es20490446e** [comment]
`ln -s foo dosbox`

I personally like names which are simple, meaningful, and easy to pronounce verbally.

`dosbox-fusion` isn't a bad option, though.

**Comment by rderooy** [comment]
How about neobox?

**Comment by krcroft** [comment]
> How about neobox?

**Comment by krcroft** [comment]
> How about neobox?

9

**Comment by es20490446e** [comment]
Sounds good too.

Perhaps we should keep a reference to the fact it's a Dos emulator on the name.

Perhaps NeoDos?

**Comment by JoshuaFern** [comment]
How about DOSRocks or DOSRox?

**Comment by dreamer** [comment]
Do we have a volunteer, who would like to have a quick look at new commits landing in dosbox-staging, and submit changes acceptable to upstream for inclusion in SVN?

If we won't have designated person for upstreaming changes by March 29th, I will update GitHub repo tagline and adjust README.md file to reflect the reality.

**Comment by krcroft** [comment]
Wanted to drop a note regarding names.. 
 - _Neo_ unfortunately gets a downvote for me because this (historically progressive) adjective has been tarnished to imply ignorant extremisms, as a label on extremist groups: Neo-liberals. Neo-nazis.  Neo-Feminists. Neo-anythingism.  We're the opposite of closed and insular here in staging; so its not a good look right now.
- In general, this repo represents "everything SVN + plus modern features, practices, and coding standards".  So the `-staging` really speaks to me; coming from the Linux world.  @dreamer's dosbox++ also has the same feel. Same with `-next`, `-ng`, `-current`, ...

**Comment by es20490446e** [comment]
Feminazi? Just block her and that's all! 😆

Really, who has time to worry about crappy people? Embrace the simple fact: THEY DON'T WANT TO CHANGE. So without knowing about them, live a long and productive life.

And that shall also reflects my thoughts on volunteering for upstream inclusion.

Neo is for NeoGeo, and the guy on the Matrix who kicks butt when someone is being naughty.

For getting an idea of a preferred option, this could come handy:
https://civs.cs.cornell.edu/

**Comment by grapeli** [comment]
@es20490446e 
One more video showing what is very slow rendering in Zone Raiders. This is best seen when the cockpit appears, which in this extreme case consists of three stages in higher resolution.
All running in dosemu-1.3.4 under [JSLinux](https://bellard.org/jslinux/vm.html?url=buildroot-x86-xwin.cfg&graphic=1) (Firefox, about 35MIPS).

https://drive.google.com/open?id=1pE4GhartUpk5dV3BlhnHlCztocdJt8YY

**Comment by es20490446e** [comment]
@grapeli But what's relevant about that? 🤔

**Comment by kcgen** [comment]
Hi @grapeli , regarding your video showing poor performance in dosemu:

> cockpit appears, which in this extreme case consists of three stages in higher resolution.
> All running in dosemu-1.3.4 under JSLinux (Firefox, about 35MIPS).

Can you take this up dosemu project?

If you notice dosbox-staging performs worse that DOSBox (0.73.x or SVN), feel free to open a new issue! Much thanks.

**Comment by es20490446e** [comment]
I mean JSLinux is a virtual machine itself. That is Dosemu there is a virtual machine running inside a virtual machine, natural it's slow. For me is incredibly slow here.

**Comment by dreamer** [comment]
I transferred the repository to newly-created https://github.com/dosbox-staging organization - this gives maintainers few more options, that were not available for a private account (and improves transparency a little bit, I hope).

All old links and references should automatically redirect to the new URL; so far transfer seems to be smooth, but if any problem will pop-up, let me know.

**Comment by kcgen** [comment]
Closing per the suggestion a couple messages up. Let's spin off related topics from here.
